scripts/data/gen-nodetime/src/relayer/lib/relayer.ts

```diff
@@ -0,0 +1,561 @@
+		// determine a unique path name from chain ids with incremental numbers. e.g.:
+		// - src-dst
+		// - src-dst-2
+		let pathName = `${srcID}-${dstID}`;
+		let suffix = "";
+		let i = 2;
+		try {
+			while (this.getPath([pathName + suffix])) {
+				suffix = `-${i}`;
+				i++;
+			}
+		} catch (e) {
+			pathName = pathName + suffix;
+		}
+
+		// construct path object and add to config.
+		try {
+			let path = {
+				id: pathName,
+				isLinked: false,
+				src: {
+					chainID: srcID,
+					portID: options.sourcePort,
+				},
+				dst: {
+					chainID: dstID,
+					portID: options.targetPort,
+				},
+			};
+
+			if (!this.config.paths)
+				this.config.paths = [];
+
+			this.config.paths.push({ path, options });
+
+			return path;
+		} catch (e) {
+			throw Errors.PathSetupFailed(e);
+		}
+	}
+
+	public getPath([id]: [string]): Path {
+		if (this.config.paths) {
+			let path = this.config.paths.find((x) => x.path.id == id);
+			if (path) return path.path;
+		}
+
+		throw Errors.PathNotExists;
+	}
+
+	public listPaths(): Path[] {
+		if (this.config.paths) {
+			let paths = this.config.paths.map((x) => x.path);
+			return paths;
+		}
+
+		throw Errors.PathsNotDefined;
+	}
+
+	public async getDefaultAccount([chainID]: [string]): Promise<Account> {
+		const chain = this.chainById(chainID);
+		if (chain) {
+			let client = await this.getIBCClient(chain);
+			return {
+				address: client.senderAddress,
+			};
+		}
+
+		throw Errors.ChainNotFound(chainID);
+	}
+
+	public async getDefaultAccountBalance([chainID]: [string]): Promise<Coin[]> {
+		const chain = this.chainById(chainID);
+		if (chain) {
+			let client = await this.getIBCClient(chain);
+			return await client.query.bank.allBalances(client.senderAddress);
+		}
+
+		throw Errors.ChainNotFound(chainID);
+	}
+
+	public async link([paths]: [string[]]): Promise<LinkResponse> {
+		if (!this.config.paths)
+			throw Errors.PathsNotDefined;
+
+		let response: LinkResponse = {
+			linkedPaths: [],
+			alreadyLinkedPaths: [],
+			failedToLinkPaths: [],
+		};
+
+		const results = [];
+
+		for (let pathName of paths) {
+			const path = this.pathById(pathName);
+
+			if (path?.path.isLinked) {
+				response.alreadyLinkedPaths.push(pathName);
+				continue;
+			}
+
+			results.push(await this.createLink(path));
+		}
+
+		for (let result of results) {
+			if (result.status) {
+				response.linkedPaths.push(result.pathName);
+			} else {
+				response.failedToLinkPaths.push({
+					pathName: result.pathName,
+					error: result.error,
+				});
+			}
+		}
+
+		return response;
+	}
+
+	public async start([paths]: [string[]]): Promise<StartResponse> {
+		if (!this.config.paths)
+			throw Errors.PathsNotDefined;
+
+		for (let pathName of paths) {
+			const path = this.pathById(pathName);
+
+			if (path?.path.isLinked) {
+				const link = await this.getLink(path);
+				setInterval(async () => {
+					let heights = this.pathById(pathName).relayerData;
+					let newHeights = await this.relayPackets(link, heights);
```

Comment:
@clockworkgr does calling this function mulltiple times at the same time frame for the same link introduce a race condition? If that's the case we need to make sure that previous call is complated before calling it again.